psydac/api/tests/test_api_feec_2d.py

```diff
@@ -0,0 +1,659 @@
+        ax1.plot(*gridlines_x1, color='k')
+        ax1.plot(*gridlines_x2, color='k')
+        ax1.set_title('Mapped grid of {} x {} cells'.format(ncells, ncells))
+        ax1.set_xlabel('x', fontsize=14)
+        ax1.set_ylabel('y', fontsize=14)
+        ax1.set_aspect('equal')
+        fig1.tight_layout()
+        fig1.show()
+
+        # ...
+        # Plot initial conditions
+        # TODO: improve
+        for i, x1i in enumerate(x1[:, 0]):
+            for j, x2j in enumerate(x2[0, :]):
+
+                Ex_values[i, j], Ey_values[i, j] = \
+                        push_2d_hcurl(E.fields[0], E.fields[1], x1i, x2j, mapping)
+
+                Bz_values[i, j] = push_2d_l2(B, x1i, x2j, mapping)
+
+        # Electric field, x component
+        fig2, ax2 = plot_field_and_error(r'E^x', x, y, Ex_values, Ex_ex(0, x, y), *gridlines)
+        fig2.show()                                             
+                                                                
+        # Electric field, y component                           
+        fig3, ax3 = plot_field_and_error(r'E^y', x, y, Ey_values, Ey_ex(0, x, y), *gridlines)
+        fig3.show()                                             
+                                                                
+        # Magnetic field, z component                           
+        fig4, ax4 = plot_field_and_error(r'B^z', x, y, Bz_values, Bz_ex(0, x, y), *gridlines)
+        fig4.show()
+        # ...
+
+        input('\nSimulation setup done... press any key to start')
+
+    # Prepare diagnostics
+    if diagnostics_interval:
+
+        # Exact energy at t=0
+        We_ex, Wb_ex = exact_energies(t)
+        diagnostics_ex['time'].append(t)
+        diagnostics_ex['electric_energy'].append(We_ex)
+        diagnostics_ex['magnetic_energy'].append(Wb_ex)
+
+        # Discrete energy at t=0
+        We_num, Wb_num = discrete_energies(e, b)
+        diagnostics_num['time'].append(t)
+        diagnostics_num['electric_energy'].append(We_num)
+        diagnostics_num['magnetic_energy'].append(Wb_num)
+
+        print('\nTotal energy in domain:')
+        print('t = {:8.4f},  exact = {Wt_ex:.13e},  discrete = {Wt_num:.13e}'.format(t,
+            Wt_ex  = We_ex  + Wb_ex,
+            Wt_num = We_num + Wb_num)
+        )
+
+    ##############################
+    print()
+    print(type(e))
+    print(type(b))
+    print()
+    print(type(M1))
+    print(type(M2))
+    print()
+    print(type(D1))
+    print(type(D1_T))
+    ##############################
+
+    #--------------------------------------------------------------------------
+    # Solution
+    #--------------------------------------------------------------------------
+
+    # TODO: add option to convert to scipy sparse format
+
+    # Arguments for time stepping
+    if periodic:
+        args = (e, b, M1, M2, D1, D1_T)
+    else:
+        args = (e, b, M1 + M1_bc, M2, D1, D1_T)
```

Comment:
the tangential trace mass matrix should probably be multiplied by a large penalization number

Reply:
Thanks for noticing that! Done in commit b2023c9